esp-hal/src/rsa/mod.rs

```diff
@@ -373,49 +372,61 @@ pub(crate) mod asynch {
     use core::task::Poll;
 
     use embassy_sync::waitqueue::AtomicWaker;
+    use portable_atomic::{AtomicBool, Ordering};
     use procmacros::handler;
 
-    use crate::{Async, rsa::{
-        Multi,Rsa,
-        RsaMode,
-        RsaModularExponentiation,
-        RsaModularMultiplication,
-        RsaMultiplication,
-    }};
+    use crate::{
+        rsa::{
+            Multi,
+            Rsa,
+            RsaMode,
+            RsaModularExponentiation,
+            RsaModularMultiplication,
+            RsaMultiplication,
+        },
+        Async,
+    };
 
     static WAKER: AtomicWaker = AtomicWaker::new();
 
+    static SIGNALED: AtomicBool = AtomicBool::new(false);
+
     /// `Future` that waits for the RSA operation to complete.
     #[must_use = "futures do nothing unless you `.await` or poll them"]
-    pub(crate) struct RsaFuture<'d> {
-        instance: &'d crate::peripherals::RSA,
+    struct RsaFuture<'a, 'd> {
+        #[cfg_attr(esp32, allow(dead_code))]
+        instance: &'a Rsa<'d, Async>,
     }
 
-    impl<'d> RsaFuture<'d> {
-        /// Asynchronously initializes the RSA peripheral.
-        fn new(instance: &'d crate::peripherals::RSA) -> Self {
+    impl<'a, 'd> RsaFuture<'a, 'd> {
+        fn new(instance: &'a Rsa<'d, Async>) -> Self {
+            SIGNALED.store(false, Ordering::Relaxed);
```

Comment:
Why not use `INT_RAW` here instead of a separate flag?

Reply:
What's that?

Reply:
So every interrupt has 4 registers. INT_ENA, INT_ST, INT_CLR and INT_RAW.

`INT_ST = INT_RAW & INT_ENA`

Reply:
I'm not sure what you're referring to, the CPU interrupt, or the peripheral. The peripherals definitely don't have any of these.

Reply:
Oh apologies 😬, I just realized this peripheral doesn't have the usual interrupt bit setup as most of the other peripherals. Ignore this comment

Reply:
Yeah this is a huge pain point with these, there are 3 iterations of the peripheral, all of them annoying 🫠

Reply:
Ideally there would normally be a way to check if the interrupt is set. Perhaps checking for idle?

Reply:
I can try some things, but not with today's brain. Yesterdays I've seen cases where, on the S3, the interrupt seemed to just disable itself for no good reason, without firing. So... I'm somewhat hesitant to touch this code right now.